File: api/src/main/java/de/storyteller/api/service/book/BookService.java
package de.storyteller.api.service.book;

import de.storyteller.api.v1.dto.book.AddBookRequest;
import de.storyteller.api.v1.dto.book.BookDTO;
import de.storyteller.api.v1.dto.book.EditBookRequest;
import de.storyteller.api.v1.dto.chapter.ChapterDTO;
import de.storyteller.api.v1.dto.cover.CoverUriDTO;

import java.util.List;
import java.util.Optional;

/**
 * Service for books
 */
public interface BookService {

    /**
     * Get all books
     * @return a list of all books
     */
    List<BookDTO> getAllBooks();

    /**
     * Get a book by its id
     * @param id the id of the book
     * @return the book with the given id
     */
    Optional<BookDTO> getBookById(String id);

    /**
     * Create a book
     * @param book the book to create
     * @return the created book
     */
    BookDTO createBook(AddBookRequest book);

    /**
     * Update a book
     * @param book the book to update
     * @return the updated book
     */
    BookDTO updateBook(EditBookRequest book);

    /**
     * Get all chapters of a book
     * @param bookId the id of the book
     * @return a list of all chapters of the book
     */
    List<ChapterDTO> getAllChapters(String bookId);

    /**
     * Update the cover of a book
     * @param id the id of the book
     * @param coverUriDTO the new cover uri
     */
    void updateBookCover(String id, CoverUriDTO coverUriDTO);

    /**
     * Get all books of an author
     * @return a list of all books of an author
     */
    List<BookDTO> getBooksByAuthor();

    /**
     * Get a book with only the published chapters
     * @param bookId the id of the book
     * @return the book with only the published chapters
     */
    BookDTO getBookWithPublishedChapters(String bookId);

    /**
     * Increase the likes of a book
     * @param bookId the id of the book
     */
    void increaseBookLikes(String bookId);

    /**
     * Decrease the likes of a book
     * @param bookId the id of the book
     */
    void decreaseBookLikes(String bookId);

    /**
     * Get the progress of a book
     * @param bookId the id of the book to get the progress of
     * @return the progress of the book
     */
    int getBookProgress(String bookId);

    /**
     * Increase the progress of a book
     * @param bookId the id of the book
     * @param progress the progress to increase
     */
    void increaseProgress(String bookId, int progress);
}

File: webapp/storyteller/src/app/service/book.service.ts
import { Injectable } from '@angular/core';
import {AddBookRequest, Book, EditBookRequest} from "../model/book";
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {environment} from "../../environments/environment";

@Injectable({
  providedIn: 'root'
})

/**
 * BookService is used to interact with the backend to get book data
 */
export class BookService {

  constructor( private http: HttpClient) { }

  /**
   * Get a book by its id
   * @param bookId the id of the book
   */
  getBookById(bookId: string): Observable<Book> {
    return this.http.get<Book>(`${environment.apiUrl}/books/${bookId}`);
  }

  /**
   * Get all books from the backend
   */
  getBooks(): Observable<Book[]> {
    return this.http.get<Book[]>(`${environment.apiUrl}/books/all`);
  }

  /**
   * Add a new book to the backend
   * @param book the book to be added
   */
  addBook(book: AddBookRequest): Observable<Book> {
    return this.http.post<Book>(`${environment.apiUrl}/books/add`, book);
  }

  /**
   * Delete a book from the backend
   */
  generateData(): Observable<any> {
    return this.http.get<any>(`${environment.apiUrl}/mock/gen`);
  }

  /**
   * Update a book in the backend
   * @param editBook the book to be updated
   */
  updateBook(editBook: EditBookRequest): Observable<EditBookRequest> {
    return this.http.put<EditBookRequest>(`${environment.apiUrl}/books/update`, editBook);
  }

  /**
   * Gets all books by the logged in author
   */
  getBooksByAuthor(): Observable<Book[]> {
    return this.http.get<Book[]>(`${environment.apiUrl}/books/user`);
  }

  /**
   * Get all books with their published chapters
   * @param bookId the id of the book
   */
  getBookWithPublishedChapters(bookId: string): Observable<Book> {
    return this.http.get<Book>(`${environment.apiUrl}/books/${bookId}/published`);
  }

  /**
   * Get the progress of a book
   * @param bookId the id of the book
   */
  getBookProgress(bookId: string): Observable<number> {
    return this.http.get<number>(`${environment.apiUrl}/books/progress/${bookId}`);
  }

  /**
   * Increase the progress of a book
   * @param bookId the id of the book
   * @param progress the current progress of the book
   */
  increaseBookProgress(bookId: string, progress: number): Observable<number> {
    console.log("Increasing progress by" + progress);
    return this.http.put<number>(`${environment.apiUrl}/books/progress/increase/${bookId}/${progress}`, {});
  }
}
